### ISSUE TITLE
curl: (52) Empty reply from server

### ISSUE THREAD

**Issue by deepsystm** [slact/nchan#464]
Hi! Can't understand what going on.
I have 2 locations: development (local) and production (digital ocean)

When I install nchan in dev env all goes ok.
When I do all same things on production env, and try to check nginx and nchan with curl:
```
 curl -i -N -H "Connection: Upgrade" -H "Upgrade: websocket" -H "Host: localhost" -H "Origin: http://localhost/sub?id=1" -H "Sec-WebSocket-Key: 123" -H "Sec-WebSocket-Extensions: permessage-deflate" -H "Sec-Websocket-Version: 13" http://localhost/sub?id=1
```
i have response:
```
HTTP/1.1 101 Switching Protocols
Server: nginx/1.13.12
Date: Fri, 06 Jul 2018 08:36:09 GMT
Content-Length: 0
Connection: upgrade
Access-Control-Allow-Origin: http://localhost/sub?id=1
Sec-WebSocket-Accept: V5hz1RKy1V4JclILDswC1e3Fek0=
Upgrade: websocket
Sec-WebSocket-Extensions: permessage-deflate

curl: (52) Empty reply from server
```

Look like nchan work, but don't proper.

nginx.conf:
```
worker_processes  1;

events {
    worker_connections  1024;
}

http {
    include       mime.types;
    default_type  application/octet-stream;

    sendfile        on;
    keepalive_timeout  65;


    server {
        listen       80;
        server_name  localhost;

        location / {
            root   html;
            index  index.html index.htm;
        }

        #error_page  404              /404.html;

        # redirect server error pages to the static page /50x.html
        #
        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root   html;
        }

        location = /sub {
      nchan_subscriber;
      nchan_channel_id $arg_id;
    }

    location = /pub {
      nchan_publisher;

      nchan_channel_id $arg_id;
    }
    }
}
```

Nginx and Nchan versions on dev and prod envs are the same.
What it could be? Thanhks!

**Comment by deepsystm** [comment]
I run deep into issue and what I found:
error throws when I compile nginx with 2 modules: passenger (5.0, 5.1, 5.2, 5.3 was tested) and nchan,
and main feature: on local environment I have:
`3.2.0-4-486 #1 Debian 3.2.63-2+deb7u2 i686 GNU/Linux` and in this env passenger + Nchan works fine

but in production servers on digital ocean I have
`4.9.0-4-amd64 #1 SMP Debian 4.9.65-3 (2017-12-03) x86_64` and when I compile nginx with Nchan I have `curl: (52) Empty reply from server` even it without passenger

I try many versions of nginx and passenger -- no result found

And when I compile in production server with `3.16.0-5-amd64 #1 SMP Debian 3.16.51-3+deb8u1 (2018-01-08) x86_64 GNU/Linux` only nginx + Nchan all is fine, but when I plug Passenger as module and compile nginx have my error `curl: (52) Empty reply from server`

**Comment by deepsystm** [comment]
try to use different versions of gcc (4.9-7.3) but it has no effect

**Comment by deepsystm** [comment]
try to use precompiled dynamic modules

**Comment by deepsystm** [comment]
Install nginx and libnginx-mod-nchan from repo 
same error

**Comment by deepsystm** [comment]
when i try to run JS i getting the reason why websockets closed with close code 1006
u can try it on 
```http://dev.flat.fm/sub?id=1
http://dev.flat.fm/pub?id=1```

**Comment by slact** [comment]
It looks like Nchan is not receiving headers from the subscriber's request, which forces it to treat the subscriber as a long-poll request by default. After that, the request is terminated after 2 minutes of inactivity if no messages are present in the channel, or a 200 OK. This suggests that the request is being proxied to your Nginx build, and headers like `Accept`, `Upgrade`, `Connection`, `Sec-Websocket-Key` and `Sec-Websocket-Version` are not being forwarded.

You should check this by `curl`ing your production machine's subscribe url locally.

**Comment by deepsystm** [comment]
Can I watch debug info of `nchan` to watch receive it headers or not?

**Comment by slact** [comment]
No, I don't think there's a way to log all the incoming headers from inside Nginx. The easiest way if to run a simple dummy HTTP server in the scripting language of your choice that dumps the entire request to stdout.

**Comment by deepsystm** [comment]
About passenger: I try to reproduce bug without passenger on nginx+nchan only and its happens on Debian 4.9 kernel05:07, July 10, 2018, slact <notifications@github.com>:No, I don't think there's a way to log all the incoming headers from inside Nginx. The easiest way if to run a simple dummy HTTP server in the scripting language of your choice that dumps the entire request to stdout.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.

-- Sent from Yandex.Mail for mobile

**Comment by deepsystm** [comment]
nothing happen? @slact

**Comment by slact** [comment]
Sorry, I haven't been able to reproduce this. If you'd like, I can look at your server. Email me at support@nchan.io to set that up.
